### PR TITLE
[Train] Implement lazy checkpointing for same-node case

### LINKED ISSUE #28362: [AIR] Replace `_TrackedCheckpoint` special case in `HuggingFaceTrainer` with a more generic system
### Description

We shouldn't need a special case there.

### Use case

_No response_

### PR DESCRIPTION
**Pull request by Yard1**
Signed-off-by: Antoni Baum <antoni.baum@protonmail.com>

## Why are these changes needed?

This PR implements lazy checkpointing for Train in the case that the worker is on the same node as Trainable, and the checkpoint is backed by a local directory. 

### Overview

Instead of serializing the Checkpoint object which happens when it is passed from worker to Trainable in a `session.report` call, the Checkpoint metadata and the path to the directory are sent over to the Trainable, and the Checkpoint is recreated in there. This saves runtime, memory and disk space. This behavior can be disabled with an environment variable.

Furthermore, the Checkpoint data is moved from the working directory to the Trial directory instead of being copied, which saves disk space. This behavior can be disabled with an environment variable.

This behavior was previously a special case in the `HuggingFaceTrainer`. That special case is now removed in this PR.

Additionally, some legacy code paths are cleaned up, and the logic in the Checkpoint object responsible for comparing paths has been made more robust.

### Notes on implementation

In order to facilitate this behavior, `_TrackedCheckpoint` can now take in an AIR Checkpoint object.

Furthermore, the default Checkpoint now has a single metadata field, which is set by Train's `TuneCheckpointManager`. This was the cleanest way of storing that metadata (previously it was always stored as a key in a dictionary checkpoint, but that required the serialization of a directory checkpoint).

Release test run: https://buildkite.com/ray-project/release-tests-pr/builds/19797

## Related issue number

Closes https://github.com/ray-project/ray/issues/28362

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by krfricke — COMMENTED** [review]
It looks to me the main shortcoming of the current checkpoint API is that `to_directory()` implies a copy, but we sometimes want it to be a move.

Could we explore other possibilities here - e.g. can we try a hardlink? And if it fails (because different disk) fallback to copy?
Are there other use cases where we (or a user) would like to move? Or should this actually be the default?

Where are the checkpoints currently stored (e.g. in huggingface trainer) and where do we want to move them?

**Comment by Yard1** [comment]
@krfricke Maybe we can make this an internal-only behavior. I think that copying is the safest and should be the default way for a user to interact with the checkpoint. Moving instead of copying could then be only used internally in specific cases.

File: python/ray/air/_internal/checkpoint_manager.py [hunk 1]
```diff
@@ -71,15 +73,21 @@ def __init__(
 
         self.metrics = flatten_dict(metrics) if metrics else {}
         self.node_ip = node_ip or self.metrics.get(NODE_IP, None)
+        # If True, and the checkpoint is an AIR Checkpoint backed by
+        # a local directory, will move files instead of copying them
+        # when commiting to disk.
+        self.move_instead_of_copy = not bool(
```
Comment (krfricke) [inline python/ray/air/_internal/checkpoint_manager.py:79]: Make internal

File: python/ray/train/_internal/backend_executor.py [hunk 2]
```diff
@@ -361,6 +363,7 @@ def initialize_session(
                     checkpoint=checkpoint,
                     encode_data_fn=encode_data_fn,
                     detailed_autofilled_metrics=use_detailed_autofilled_metrics,
+                    enable_lazy_checkpointing=use_lazy_checkponting,
```
Comment (Yard1) [inline python/ray/train/_internal/backend_executor.py:366]:
```suggestion
                    enable_lazy_checkpointing=use_lazy_checkpointing,
```

**Comment by amogkam** [comment]
Overall lgtm. Will leave to @krfricke for final review and approval/merging/

File: python/ray/air/tests/test_checkpoints.py [hunk 3]
```diff
@@ -127,6 +128,20 @@ def test_directory(self):
...
+        assert "foo" in checkpoint._SERIALIZED_ATTRS
+        checkpoint.foo = "bar"
+
+        path = checkpoint.to_directory()
+        recovered_checkpoint = StubCheckpoint.from_directory(path)
+        tmpdir = tempfile.mkdtemp()
+        new_path = recovered_checkpoint._move_directory(tmpdir)
+        new_recovered_checkpoint = StubCheckpoint.from_directory(new_path)
+
+        assert new_recovered_checkpoint.foo == "bar"
```
Comment (krfricke) [inline python/ray/air/tests/test_checkpoints.py:142]:
```suggestion
        assert recovered_checkpoint._local_path == tmpdir
        assert new_recovered_checkpoint.foo == "bar"
```

**Review by krfricke — APPROVED** [review]
Lokks great!

File: python/ray/train/tests/test_data_parallel_trainer_checkpointing.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    ("dir", False),
+    ("lazy_dir", False),
+)
+
+
+@pytest.mark.parametrize(
+    "checkpoint_type_and_should_copy", checkpoint_type_and_should_copy
+)
+def test_checkpoint(ray_start_4_cpus, checkpoint_type_and_should_copy):
+    checkpoint_type, should_copy = checkpoint_type_and_should_copy
```
Comment (krfricke) [inline python/ray/train/tests/test_data_parallel_trainer_checkpointing.py:55]: I know most of these tests are copied over, but can we add a docstring to each of these to describe on a high-level what we are testing here?

Doesn't have to be very verbose. Example:

https://github.com/ray-project/ray/pull/30777/files#diff-6bd76b52e9384d1db40c10754844dd75a07dcfed90dc4b0e69c8ee9dcc720af8R63

**Comment by Yard1** [comment]
Marking as WIP until https://github.com/ray-project/ray/pull/30948 is merged

**Comment by krfricke** [comment]
Can you merge latest master again to kick off tests (there was a github issue)?

**Pull request merged**